packages/fleet/lib/supervisor.ts

```diff
@@ -0,0 +1,278 @@
+    | { type: 'OUTDATE'; node: Node }
+    | { type: 'TERMINATE'; node: Node }
+    | { type: 'REMOVE'; node: Node }
+    | { type: 'WARN'; node: Node };
+
+type SupervisorState = 'stopped' | 'running' | 'stopping';
+
+export class Supervisor {
+    private state: SupervisorState = 'stopped';
+    private dbClient: DatabaseClient;
+    private nodeProvider: NodeProvider;
+    public STATE_TIMEOUT_MS = {
+        STARTING: 5 * 60 * 1000,
+        FINISHING: 24 * 60 * 60 * 1000,
+        TERMINATED: 7 * 24 * 60 * 60 * 1000,
+        ERROR: 7 * 24 * 60 * 60 * 1000
+    };
+
+    constructor({ dbClient, nodeProvider }: { dbClient: DatabaseClient; nodeProvider: NodeProvider }) {
+        this.dbClient = dbClient;
+        this.nodeProvider = nodeProvider;
+    }
+
+    public async start(): Promise<void> {
+        if (this.state === 'running') {
+            logger.warn('Supervisor is already running');
+            return;
+        }
+
+        this.state = 'running';
+        await this.loop();
+    }
+
+    public async stop(): Promise<void> {
+        this.state = 'stopping';
+        logger.info('Supervisor stopped');
+
+        // wait for the loop to finish or timeout
+        const waitForStopped = async () => {
+            while (this.state !== 'stopped') {
+                await setTimeout(100);
+            }
+        };
+        await Promise.race([waitForStopped(), setTimeout(60000)]);
+    }
+
+    public async tick(): Promise<void> {
+        // TODO: trace
+        try {
+            const plan = await this.plan();
+            if (plan.isOk()) {
+                await this.executePlan(plan.value);
+            } else {
+                logger.error('Supervision error:', plan.error);
+                await setTimeout(1000);
+            }
+        } catch (error) {
+            logger.error('Supervision error:', error);
+        }
+    }
+
+    private async loop(): Promise<void> {
+        while (this.state === 'running') {
+            // TODO: Lock to prevent multiple instances from running
+            await this.tick();
+        }
+        this.state = 'stopped';
+    }
+
+    private async plan(cursor?: number): Promise<Result<Action[]>> {
+        const getDeployment = await deployments.getActive(this.dbClient.db);
+        if (getDeployment.isErr()) {
+            return Err(getDeployment.error);
+        }
+        if (!getDeployment.value) {
+            return Err(new FleetError('no_active_deployment'));
+        }
+        const deployment = getDeployment.value;
+        const plan: Action[] = [];
+
+        const search = await nodes.search(this.dbClient.db, {
+            states: ['PENDING', 'STARTING', 'RUNNING', 'OUTDATED', 'FINISHING', 'IDLE', 'TERMINATED', 'ERROR'],
+            ...(cursor ? { cursor } : {})
+        });
+        if (search.isErr()) {
+            return Err(search.error);
+        }
+        for (const [routingId, nodes] of search.value.nodes) {
```

Comment:
this part is confusing to me, why would group nodes by `routingId`, what's the use? 
Not grouping would make the loop look more regular, e.g: `if (starting) else if (error) etc.`

Reply:
because the decision to create a new node if there is no RUNNING or upcoming ones is per `routingId`. 
True that the rest of the decisions are not dependent on the `routingId` though.